clients/branch.go

```diff
@@ -16,9 +16,10 @@ package clients
 
 // BranchRef represents a single branch reference and its protection rules.
 type BranchRef struct {
-	Name                 *string
-	Protected            *bool
-	BranchProtectionRule BranchProtectionRule
+	Name                     *string
+	Protected                *bool
+	WereAllSettingsAvailable *bool
```

Comment:
my hesitation to put it here is that this is shared with gitlab, and really this quirk is with GitHub

Reply:
So the problem is we don't know if we can trust a 0 for required review count when we get a ref update rule right?

So can we just clear the field if its 0?
https://github.com/ossf/scorecard/blob/74c57cd1bc54d643a5e77fa4c3c3c2c1309ff84f/clients/githubrepo/branches.go#L356-L363

Reply:
That is an interesting idea, but we'd need to think on what "clear the field" would mean, because the `nil` value on the `RequiredPullRequestReviews.RequiredApprovingReviewCount` already means "no PR is required". But I think setting the whole `RequiredPullRequestReviews` pointer to `nil` would work, because if Scorecard is run without admin, all the other information inside this structure (dismissStaleReviews and RequireCodeOwners) is coming with `nil`. Does that work for you?

Reply:
I was talking about setting `RequiredApprovingReviewCount` to `nil`, as it's the minimal required change to do what the PR is trying to do.

Reply:
Yes, but I don't think this would work, because if we set `RequiredApprovingReviewCount` to `nil`, the evaluation part would interpret that no PRs are required (that's how GitHub responses register it), but that is not true. In that case, we actually can't say if it's not required or it's required without any reviewers required.

Reply:
Currently Tier 2 gives you 3 points (6/10 total) if you require 1+ reviewer, otherwise you're stuck with 3/10 from Tier 1.

With a non-admin token, (since you need admin to see if PRs are required but 0 reviewers), you will still score a 3/10 as its impossible to distinguish between PRs required and no PRs required. So clearing `RequiredApprovingReviewCount ` here is fine.

With an admin token, we won't clear `RequiredApprovingReviewCount` because we'll have access to `branchProtectionRule` and we're only clearing the field for the `refUpdateRule` case.

Reply:
This "clear" of the `RequiredPullRequestReviews` would happen only for the case of Branch-Protection run without admin token, which is a case that we shouldn't explore nor change much in the future, given that GitHub publicly exposed that they plan to move to Repo Rules. At my point of view, this also make this change less impactful, as it's a corner case that will have less and less occurrence over time.

Reply:
Hmm I see your point, I think it makes sense

Reply:
I guess the drawback from this decision is that the score for non-admin users would probably become inconsistent if we decide to remove the tier scoring (#3123) or have any big changes on the current tier requirements. Does it makes sense?

Reply:
> I'm assuming "this decision" is referring to the whole PR which recognizes PRs are required, not just which pointer to clear? If so, yes I see what you're saying.

Although I agree with your points about scoring with many data points and weighed approach, I actually meant something else, I'll try to explain better.

From this conversation, I'm seeing two main paths on how to score the "Require PRs prior to make any code changes" requirement on the specific case of Scorecard run without admin token on a repo that uses old Branch-Protection rules (case where we can't really know if the require PRs or not):
1. We set the `RequiredPullRequestReviews.RequiredApprovingReviewCount` to `nil`, forcing our Evaluation part of the code to interpret that the repository does not require PRs.
2. Make some different change on the data that goes to the Evaluation part to mean "we don't know if it requires PRs or not". This would be the goal of my approach that sent `WereAllSettingsAvailable` on the branch structure, or the one that sent the whole `RequiredPullRequestReviews` to nil.

What I meant with the "this decision" was the case described on 1, because it would force the negative outcome of the "Require PRs prior to make any code changes" requirement -- this does not impact our current scoring, because the "Require 1 reviewer" rule is on the same tier and would block the score anyway. 
But for the approaches on 2, the requirement wouldn't have a negative outcome, but would be managed as a admin-only requirement that shouldn't block the maintainer to reach scores of higher tiers, nor affect negatively their score in case we remove tiering.

Reply:
That said, I have one other approach to suggest, that also conveys with the goal explained on the 2nd point of my previous comment: 
We could set a "flag value" like `-1` on the `RequiredPullRequestReviews.RequiredApprovingReviewCount` to mean that we don't have the information. I know that this information is usually given as a `nil` value, but GitHub's API already uses `nil` to say that no PR is required, so we can't use this same value. This would be an internal interpretation, understood and followed only between our GitHub Branch Protection client and the Branch Protection part of our Evaluation code

This approach would avoid having a negative outcome on a requirement we actually can't evaluate, but also would avoid unnecessary changes on our data structures (as my first suggestion currently implemented on this PR) or clearing the pointer of the whole `RequiredPullRequestReviews` structure, which could lead to unwanted outcomes.

Let me know what you think.

Reply:
> We could set a "flag value" like -1 on the RequiredPullRequestReviews.RequiredApprovingReviewCount to mean that we don't have the information. 

In my opinion, we already have the "flag value" convention, it's a `nil` pointer. Happy to have a chat about this this week to clear confusion on either side.

Reply:
We have access to two datapoints tell us if they use PRs:
- more than 1 required reviewer
- require codeowner

If we dont have either, then clear the whole `RequiredPullRequestReviews` pointer, as we dont have any data. This logic should live in the code block highlighted above:
https://github.com/ossf/scorecard/blob/74c57cd1bc54d643a5e77fa4c3c3c2c1309ff84f/clients/githubrepo/branches.go#L356-L363